### PR TITLE
[full-ci] Add inmem option for nats-js stores

### PR DESCRIPTION
**Pull request by kobergj**
Adds an option to disable persistence of nats-js and nats-js-kv stores

Fixes https://github.com/owncloud/ocis/issues/7119

### CODE REVIEW

File: services/gateway/README.md [hunk 1]
```diff
@@ -23,3 +23,4 @@ Note: The gateway service can only be scaled if not using `memory` store and the
 Store specific notes:
   -   When using `redis-sentinel`, the Redis master to use is configured via e.g. `OCIS_CACHE_STORE_NODES` in the form of `<sentinel-host>:<sentinel-port>/<redis-master>` like `10.10.0.200:26379/mymaster`.
   -   When using `nats-js-kv` it is recommended to set `OCIS_CACHE_STORE_NODES` to the same value as `OCIS_EVENTS_ENDPOINT`. That way the cache uses the same nats instance as the event bus.
+  -   When using `nats-js-kv` store it is possible to use `OCIS_CACHE_DISABLE_PERSISTENCE` to instruct nats to not persist cache entries on disc.
```
Comment (mmattel) [inline services/gateway/README.md:26]:
```suggestion
  -   When using the `nats-js-kv` store, it is possible to set `OCIS_CACHE_DISABLE_PERSISTENCE` to instruct nats to not persist cache data on disc.
```

File: services/frontend/pkg/config/config.go [hunk 2]
```diff
@@ -126,16 +126,18 @@ type DataGateway struct {
...
-	StatCacheSize                        int                `yaml:"stat_cache_size" env:"OCIS_CACHE_SIZE;FRONTEND_OCS_STAT_CACHE_SIZE" desc:"Max number of entries to hold in the cache."`
+	Prefix                      string        `yaml:"prefix" env:"FRONTEND_OCS_PREFIX" desc:"URL path prefix for the OCS service. Note that the string must not start with '/'."`
+	SharePrefix                 string        `yaml:"share_prefix" env:"FRONTEND_OCS_SHARE_PREFIX" desc:"Path prefix for shares as part of an ocis resource. Note that the path must start with '/'."`
+	HomeNamespace               string        `yaml:"home_namespace" env:"FRONTEND_OCS_PERSONAL_NAMESPACE" desc:"Homespace namespace identifier."`
+	AdditionalInfoAttribute     string        `yaml:"additional_info_attribute" env:"FRONTEND_OCS_ADDITIONAL_INFO_ATTRIBUTE" desc:"Additional information attribute for the user like {{.Mail}}."`
+	StatCacheType               string        `yaml:"stat_cache_type" env:"OCIS_CACHE_STORE;FRONTEND_OCS_STAT_CACHE_STORE" desc:"The type of the cache store. Supported values are: 'memory', 'ocmem', 'etcd', 'redis', 'redis-sentinel', 'nats-js', 'noop'. See the text description for details."`
```
Comment (mmattel) [inline services/frontend/pkg/config/config.go:133]: Is'nt `nats-js-kv` missing in the list of stores?
It is referenced via `OCIS_CACHE_DISABLE_PERSISTENCE` but not mentioned.
If missing:
* add to all `OCIS_CACHE_STORE` occurrences
* fix the readme of the corresponding services to add `nats-js-kv` as possible item to the cache list

Comment (kobergj) [reply services/frontend/pkg/config/config.go]: Done. Could you have another look?

File: services/eventhistory/README.md [hunk 3]
```diff
@@ -14,17 +14,22 @@ The `eventhistory` services consumes all events from the configured event system
...
+  -   `nats-js`: Stores data using object-store feature of [nats jetstream](https://docs.nats.io/nats-concepts/jetstream/obj_store) (deprecated)
 
-1.  Note that in-memory stores are by nature not reboot-persistent.
-2.  Though usually not necessary, a database name and a database table can be configured for event stores if the event store supports this. Generally not applicable for stores of type `in-memory`. These settings are blank by default which means that the standard settings of the configured store apply.
-4.  The eventhistory service can be scaled if not using `in-memory` stores and the stores are configured identically over all instances.
-5.  When using `redis-sentinel`, the Redis master to use is configured via `EVENTHISTORY_STORE_NODES` in the form of `<sentinel-host>:<sentinel-port>/<redis-master>` like `10.10.0.200:26379/mymaster`.
+Other store types may work but are not supported currently.
+
+Note: The service can only be scaled if not using `memory` store and the stores are configured identically over all instances!
+
```
Comment (mmattel) [inline services/eventhistory/README.md:28]:
```suggestion

Note that if you have used one of the deprecated stores, you should reconfigure to one of the supported ones as the deprecated stores will be removed in a later version.

```

File: services/frontend/README.md [hunk 4]
```diff
@@ -44,17 +44,22 @@ You can find more details regarding available attributes at the [libre-graph-api
...
+  -   `nats-js`: Stores data using object-store feature of [nats jetstream](https://docs.nats.io/nats-concepts/jetstream/obj_store) (deprecated)
 
-1.  Note that in-memory stores are by nature not reboot-persistent.
-2.  Though usually not necessary, a database name and a database table can be configured for event stores if the event store supports this. Generally not applicable for stores of type `in-memory`. These settings are blank by default which means that the standard settings of the configured store apply.
-3.  The frontend service can be scaled if not using `in-memory` stores and the stores are configured identically over all instances.
-4.  When using `redis-sentinel`, the Redis master to use is configured via `FRONTEND_OCS_STAT_CACHE_STORE_NODES` in the form of `<sentinel-host>:<sentinel-port>/<redis-master>` like `10.10.0.200:26379/mymaster`.
+Other store types may work but are not supported currently.
+
+Note: The service can only be scaled if not using `memory` store and the stores are configured identically over all instances!
+
```
Comment (mmattel) [inline services/frontend/README.md:58]:
```suggestion

Note that if you have used one of the deprecated stores, you should reconfigure to one of the supported ones as the deprecated stores will be removed in a later version.

```

File: services/gateway/README.md [hunk 5]
```diff
@@ -15,11 +15,16 @@ Use `OCIS_CACHE_STORE` (`GATEWAY_PROVIDER_CACHE_STORE`, `GATEWAY_CREATE_HOME_CAC
...
   -   `noop`: Stores nothing. Useful for testing. Not recommended in production environments.
+  -   `ocmem`: Advanced in-memory store allowing max size. (deprecated)
+  -   `redis`: Stores data in a configured Redis cluster. (deprecated)
+  -   `etcd`: Stores data in a configured etcd cluster. (deprecated)
+  -   `nats-js`: Stores data using object-store feature of [nats jetstream](https://docs.nats.io/nats-concepts/jetstream/obj_store) (deprecated)
 
 Other store types may work but are not supported currently.
 
-Note: The gateway service can only be scaled if not using `memory` store and the stores are configured identically over all instances!
+Note: The service can only be scaled if not using `memory` store and the stores are configured identically over all instances!
```
Comment (mmattel) [inline services/gateway/README.md:26]:
```suggestion

Note that if you have used one of the deprecated stores, you should reconfigure to one of the supported ones as the deprecated stores will be removed in a later version.

```

File: services/graph/README.md [hunk 6]
```diff
@@ -19,24 +19,29 @@ The following image gives an overview of the scenario when a client requests to
...
+  -   `nats-js`: Stores data using object-store feature of [nats jetstream](https://docs.nats.io/nats-concepts/jetstream/obj_store) (deprecated)
 
-1.  Note that in-memory stores are by nature not reboot-persistent.
-2.  Though usually not necessary, a database name and a database table can be configured for event stores if the event store supports this. Generally not applicable for stores of type `in-memory`. These settings are blank by default which means that the standard settings of the configured store apply.
-3.  The graph service can be scaled if not using `in-memory` stores and the stores are configured identically over all instances.
-4.  When using `redis-sentinel`, the Redis master to use is configured via `GRAPH_CACHE_STORE_NODES` in the form of `<sentinel-host>:<sentinel-port>/<redis-master>` like `10.10.0.200:26379/mymaster`.
+Other store types may work but are not supported currently.
+
+Note: The service can only be scaled if not using `memory` store and the stores are configured identically over all instances!
+
```
Comment (mmattel) [inline services/graph/README.md:33]:
```suggestion

Note that if you have used one of the deprecated stores, you should reconfigure to one of the supported ones as the deprecated stores will be removed in a later version.

```

File: services/postprocessing/README.md [hunk 7]
```diff
@@ -20,17 +20,22 @@ The `postprocessing` service needs to store some metadata about uploads to be ab
...
+  -   `nats-js`: Stores data using object-store feature of [nats jetstream](https://docs.nats.io/nats-concepts/jetstream/obj_store) (deprecated)
 
-1.  Note that in-memory stores are by nature not reboot-persistent.
-2.  Though usually not necessary, a database name and a database table can be configured for event stores if the event store supports this. Generally not applicable for stores of type `in-memory`. These settings are blank by default which means that the standard settings of the configured store apply.
-3.  The postprocessing service can be scaled if not using `in-memory` stores and the stores are configured identically over all instances.
-4.  When using `redis-sentinel`, the Redis master to use is configured via `POSTPROCESSING_STORE_NODES` in the form of `<sentinel-host>:<sentinel-port>/<redis-master>` like `10.10.0.200:26379/mymaster`.
+Other store types may work but are not supported currently.
+
+Note: The service can only be scaled if not using `memory` store and the stores are configured identically over all instances!
+
```
Comment (mmattel) [inline services/postprocessing/README.md:34]:
```suggestion

Note that if you have used one of the deprecated stores, you should reconfigure to one of the supported ones as the deprecated stores will be removed in a later version.

```

File: services/proxy/README.md [hunk 8]
```diff
@@ -134,17 +134,22 @@ In a production deployment, you want to have basic authentication (`PROXY_ENABLE
...
+  -   `nats-js`: Stores data using object-store feature of [nats jetstream](https://docs.nats.io/nats-concepts/jetstream/obj_store) (deprecated)
 
-1.  Note that in-memory stores are by nature not reboot-persistent.
-2.  Though usually not necessary, a database name and a database table can be configured for event stores if the event store supports this. Generally not applicable for stores of type `in-memory`. These settings are blank by default which means that the standard settings of the configured store apply.
-3.  The proxy service can be scaled if not using `in-memory` stores and the stores are configured identically over all instances.
-4.  When using `redis-sentinel`, the Redis master to use is configured via `PROXY_OIDC_USERINFO_CACHE_NODES` in the form of `<sentinel-host>:<sentinel-port>/<redis-master>` like `10.10.0.200:26379/mymaster`.
+Other store types may work but are not supported currently.
+
+Note: The service can only be scaled if not using `memory` store and the stores are configured identically over all instances!
+
```
Comment (mmattel) [inline services/proxy/README.md:148]:
```suggestion

Note that if you have used one of the deprecated stores, you should reconfigure to one of the supported ones as the deprecated stores will be removed in a later version.

```

File: services/settings/README.md [hunk 9]
```diff
@@ -41,26 +41,23 @@ graph TD
...
-    stores of type `in-memory`, `redis` and `redis-sentinel`. These settings
-    are blank by default which means that the standard settings of the
-    configured store apply.
-3.  The `settings` service can be scaled if not using `in-memory` stores and
-    the stores are configured identically over all instances.
-4.  When using `redis-sentinel`, the Redis master to use is configured via
-    `SETTINGS_CACHE_STORE_NODES` in the form of
-    `<sentinel-host>:<sentinel-port>/<redis-master>` like
-    `10.10.0.200:26379/mymaster`.
+Store specific notes:
```
Comment (mmattel) [inline services/settings/README.md:57]:
```suggestion

Note that if you have used one of the deprecated stores, you should reconfigure to one of the supported ones as the deprecated stores will be removed in a later version.

Store specific notes:
```

File: services/storage-system/README.md [hunk 10]
```diff
@@ -10,13 +10,19 @@ Starting with ocis version 3.0.0, the default backend for metadata switched to m
...
+  -   `nats-js`: Stores data using object-store feature of [nats jetstream](https://docs.nats.io/nats-concepts/jetstream/obj_store) (deprecated)
 
-1.  Note that in-memory stores are by nature not reboot-persistent.
-2.  Though usually not necessary, a database name can be configured for event stores if the event store supports this. Generally not applicable for stores of type `in-memory`, `redis` and `redis-sentinel`. These settings are blank by default which means that the standard settings of the configured store apply.
-3.  The `storage-system` service can be scaled if not using `in-memory` stores and the stores are configured identically over all instances.
-4.  When using `redis-sentinel`, the Redis master to use is configured via `STORAGE_SYSTEM_CACHE_NODES` in the form of `<sentinel-host>:<sentinel-port>/<redis-master>` like `10.10.0.200:26379/mymaster`.
+Other store types may work but are not supported currently.
+
+Note: The service can only be scaled if not using `memory` store and the stores are configured identically over all instances!
+
```
Comment (mmattel) [inline services/storage-system/README.md:24]:
```suggestion

Note that if you have used one of the deprecated stores, you should reconfigure to one of the supported ones as the deprecated stores will be removed in a later version.

```

File: services/storage-users/README.md [hunk 11]
```diff
@@ -101,13 +101,19 @@ The configuration for the `purge-expired` command is done by using the following
...
+  -   `nats-js`: Stores data using object-store feature of [nats jetstream](https://docs.nats.io/nats-concepts/jetstream/obj_store) (deprecated)
 
-1.  Note that in-memory stores are by nature not reboot-persistent.
-2.  Though usually not necessary, a database name can be configured for event stores if the event store supports this. Generally not applicable for stores of type `in-memory`, `redis` and `redis-sentinel`. These settings are blank by default which means that the standard settings of the configured store apply.
-3.  The `storage-users` service can be scaled if not using `in-memory` stores and the stores are configured identically over all instances.
-4.  When using `redis-sentinel`, the Redis master to use is configured via `STORAGE_USERS_STAT_CACHE_STORE_NODES`, `STORAGE_USERS_FILEMETADATA_CACHE_STORE_NODES` and `STORAGE_USERS_ID_CACHE_STORE_NODES` in the form of `<sentinel-host>:<sentinel-port>/<redis-master>` like `10.10.0.200:26379/mymaster`.
+Other store types may work but are not supported currently.
+
+Note: The service can only be scaled if not using `memory` store and the stores are configured identically over all instances!
+
```
Comment (mmattel) [inline services/storage-users/README.md:115]:
```suggestion

Note that if you have used one of the deprecated stores, you should reconfigure to one of the supported ones as the deprecated stores will be removed in a later version.

```

File: services/userlog/README.md [hunk 12]
```diff
@@ -17,17 +17,22 @@ Running the `userlog` service without running the `eventhistory` service is not
...
+  -   `nats-js`: Stores data using object-store feature of [nats jetstream](https://docs.nats.io/nats-concepts/jetstream/obj_store) (deprecated)
 
-1.  Note that in-memory stores are by nature not reboot-persistent.
-2.  Though usually not necessary, a database name and a database table can be configured for event stores if the event store supports this. Generally not applicable for stores of type `in-memory`. These settings are blank by default which means that the standard settings of the configured store apply.
-3.  The userlog service can be scaled if not using `in-memory` stores and the stores are configured identically over all instances.
-4.  When using `redis-sentinel`, the Redis master to use is configured via `USERLOG_STORE_NODES` in the form of `<sentinel-host>:<sentinel-port>/<redis-master>` like `10.10.0.200:26379/mymaster`.
+Other store types may work but are not supported currently.
+
+Note: The service can only be scaled if not using `memory` store and the stores are configured identically over all instances!
+
```
Comment (mmattel) [inline services/userlog/README.md:31]:
```suggestion

Note that if you have used one of the deprecated stores, you should reconfigure to one of the supported ones as the deprecated stores will be removed in a later version.

```

**Review by 2403905 — APPROVED** [review]

File: go.mod [hunk 13]
```diff
@@ -350,3 +350,5 @@ require (
 replace github.com/go-micro/plugins/v4/store/nats-js-kv => github.com/kobergj/plugins/v4/store/nats-js-kv v0.0.0-20231207143248-4d424e3ae348
 
 replace github.com/studio-b12/gowebdav => github.com/aduffeck/gowebdav v0.0.0-20231123085457-ff658b6ea159
+
+replace github.com/cs3org/reva/v2 => github.com/kobergj/reva/v2 v2.0.0-20231214151714-2516d234cdc0
```
Comment (2403905) [inline go.mod:354]: Should be removed?

**Review by 2403905 — APPROVED** [review]

File: go.mod (hunk 13 above)
Comment (kobergj) [reply go.mod]: Yes ofc. Good find :+1:

**Review by mmattel — APPROVED** [review]
LGTM from a docs pov 👍

**Review by micbar — APPROVED** [review]

**Pull request merged**